students/thecthaeh/Lesson04/mailroom_part2.py

```diff
@@ -60,32 +55,36 @@ def handle_selection2():
     report = []
     
     #fill the report with donor information
-    for donor in donor_info:
-        report.append([donor[0], sum(donor[1]), len(donor[1]), sum(donor[1])/len(donor[1])])
+    for donor in donor_info_dict:
+        report.append([donor, sum(donor_info_dict[donor]), len(donor_info_dict[donor]), sum(donor_info_dict[donor])/len(donor_info_dict[donor])])
     
     #sort the report
-    sorted_report = sorted(report, key = itemgetter(1), reverse = True)
+    sorted_report = sorted(report, key=itemgetter(1), reverse=True)
     
     #print the sorted report
     for donor in sorted_report:
         print("{:30}  ${:>19.2f}  {:>15}  ${:>19.2f}".format(*donor[:]))
     
-    prompt()
+    return prompt()
 
-#This dict holds the different functions for each selection option
-select_func_dict = {
-    1: handle_selection1, 
-    2: handle_selection2,
-}
+#function for selection 3
+def thank_all():
+    for name, donation in donor_info_dict.items():
+        with open(f"./{name}.txt", 'w') as f:
+            f.write("Dear {},\n\nYou have donated a total of ${:.2f}. \n\nYour generosty will help us fulfill our plans for the coming year. We will send you updates on our upcoming projects so you can see how your donations are being used.\n\nThank you!\nThe Team\n".format(name, sum(donation)))
+    return prompt()
 
 def main(): #use a dict to switch between the user's selections
-    prompt()
-    while actions != '3':
-        if actions == '1':
-            select_func_dict.get(1)()
-            
-        elif actions == '2':
-            select_func_dict.get(2)()
+    actions = prompt()
+    while True:
+        if actions == '4' or actions == 'quit':
+            break
+        menu = {
+            1: single_thank_you, 
+            2: create_report,
+            3: thank_all
+        }
+        actions = menu.get(int(actions))()
```

Comment:
You are still calling "main" within other functions. Your "feature" functions should be unaware of the main flow.
See reference to this in the tutorial: https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html


couple other comments:
* your switch dict mapping should be outside of while loop
* promt (input()) function should be inside while loop
* quit options should be mapped to a function which calls sys.exit()


so your code should look like this:
```
menu = {"1": single_thank_you, "2": create_report, "3": thank_all, "4": quit_program, "quit": "quit_program}
while True:
    response = input("...")
    menu[response]()

```
that's it!

Reply:
Ah. So none of my functions should need to call prompt()?

Reply:
Exactly!

your prompt() or input() directly would just be once inside of a while True loop